python/lsst/pipe/base/_task_metadata.py

```diff
@@ -0,0 +1,427 @@
+
+        Parameters
+        ----------
+        key : `str`
+            Item to retrieve.
+
+        Returns
+        -------
+        value : Any
+            Either the value associated with the key or, if the key
+            corresponds to a list, the last item in the list.
+        """
+        # Used in pipe_tasks
+        keys = self._getKeys(key)
+        key0 = keys.pop(0)
+        if len(keys) == 0:
+            if key0 in self.arrays:
+                return self.arrays[key0][-1]
+            elif key0 in self.scalars:
+                return self.scalars[key0]
+            elif key0 in self.metadata:
+                return self.metadata[key0]
+            raise KeyError(f"'{key}' not found")
+
+        return self.metadata[key0].getScalar(".".join(keys))
+
+    def getArray(self, key):
+        """Retrieve an item as a list even if it is a scalar.
+
+        Parameters
+        ----------
+        key : `str`
+            Item to retrieve.
+
+        Returns
+        -------
+        values : `list` of any
+            A list containing the value or values associated with this item.
+        """
+        keys = self._getKeys(key)
+        key0 = keys.pop(0)
+        if len(keys) == 0:
+            if key0 in self.arrays:
+                return self.arrays[key0]
+            elif key0 in self.scalars:
+                return [self.scalars[key0]]
+            elif key0 in self.metadata:
+                return [self.metadata[key0]]
+            raise KeyError(f"'{key}' not found")
+
+        return self.metadata[key0].getArray(".".join(keys))
+
+    def names(self, topLevelOnly: bool = True):
+        """Return the hierarchical keys from the metadata.
+
+        Parameters
+        ----------
+        topLevelOnly  : `bool`
+            If true, return top-level keys, otherwise full metadata item keys.
+
+        Returns
+        -------
+        names : `collection.abc.Set`
+            A set of top-level keys or full metadata item keys, including
+            the top-level keys.
+
+        Notes
+        -----
+        Should never be called in new code with ``topLevelOnly`` set to `True`
+        -- this is equivalent to asking for the keys and is the default
+        when iterating through the task metadata. In this case a deprecation
+        message will be issued and the ability will raise an exception
+        in a future release.
+
+        When ``topLevelOnly`` is `False` all keys, including those from the
+        hierarchy and the top-level hierarchy, are returned.
+        """
+        if topLevelOnly:
+            warnings.warn("Use keys() instead. " + _DEPRECATION_REASON, FutureWarning)
+            return set(self.keys())
+        else:
+            names = set()
+            for k, v in self.items():
+                names.add(k)  # Always include the current level
+                if isinstance(v, TaskMetadata):
+                    names.update({k + '.' + item for item in v.names(topLevelOnly=topLevelOnly)})
```

Comment:
This is a tiny bit confusing since you know that `topLevelOnly` is `False`.

Reply:
I used the given value mostly because if I rearranged the code at some point I wouldn't need to change the value there.